lalrpop/src/build/mod.rs

```diff
@@ -266,53 +265,68 @@ fn lalrpop_files<P: AsRef<Path>>(root_dir: P) -> io::Result<Vec<PathBuf>> {
 }
 
 fn parse_and_normalize_grammar(session: &Session, file_text: &FileText) -> io::Result<r::Grammar> {
-    let grammar = match parser::parse_grammar(file_text.text()) {
-        Ok(grammar) => grammar,
+    let grammar = parser::parse_grammar(file_text.text())
+        .map_err(|error| report_parse_error(file_text, error, report_error))?;
 
-        Err(ParseError::InvalidToken { location }) => {
+    match normalize::normalize(session, grammar) {
+        Ok(grammar) => Ok(grammar),
+        Err(error) => Err(report_error(
+            file_text,
+            error.span,
+            &error.message,
+            io::Error::from(io::ErrorKind::InvalidData),
+        ))?,
+    }
+}
+
+pub fn report_parse_error<E>(
+    file_text: &FileText,
+    error: parser::ParseError<'_>,
+    mut reporter: impl FnMut(&FileText, pt::Span, &str, io::Error) -> E,
```

Comment:
This implementation where the callback takes in an io::Error and then turns around and returns that same error (but now it has the generic type E) feels weird.  It looks like in the LSP implementation it just re-wraps up the io::Error.

io::Error was always a sort of weird return here.  The only reason the block that got factored out returned an io::Error is because that's the error type that goes up the stack.  I think rather than do this whole "pass the io::Error around" dance, it would make more sense to just have the internal implementation use io::ErrorKind::InvalidData unconditionally.  I don't care too much to preserve the one UnexpectedEof case, which is only arguably important in the first place.

Reply:
Good point. If we don't need to keep the io::Error around, then we may consider removing it entirely.